pennylane/templates/state_preparations/superposition.py

```diff
@@ -0,0 +1,335 @@
+
+    .. code-block::
+
+        coeffs = np.sqrt([1/3, 1/3, 1/3])
+        basis = [[1, 1, 1], [0, 1, 0], [0, 0, 0]]
+        wires = [0, 1, 2]
+        work_wire = 3
+
+        dev = qml.device('default.qubit')
+        @qml.qnode(dev)
+        def circuit():
+            qml.Superposition(coeffs, basis, wires, work_wire)
+            return qml.probs(wires)
+
+    .. code-block:: pycon
+
+        >>> print(circuit())
+        [0.33333333 0.         0.33333333 0.         0.         0.
+        0.         0.33333333]
+
+
+    .. details::
+        :title: Details
+
+        The construction of this template is divided into two blocks.
+        The first block takes the list of coefficients :math:`c_i` and prepares the state:
+
+        .. math::
+
+            |\phi\rangle = \sum_i^m c_i |i\rangle,
+
+        where :math:`|i\rangle` are the computational basis states and :math:`m` is the number of terms
+        in the superposition. This is done using the
+        :class:`~.StatePrep` template in the fisrt :math:`\lceil \log_2 m \rceil` qubits.
+        This significantly reduces complexity by not having to work on the entire system.
+
+        The second block is responsible for the permutation of the basis states previously prepared to
+        the target basis states.
+
+        .. math::
+
+            |i\rangle \rightarrow |b_i\rangle.
+
+        This block maps the elements one by one using an auxiliary qubit.
+        The process can be divided into three operations. Let's assume
+        we want to map :math:`|i\rangle` to :math:`|b_i\rangle`:
+
+        1. By using a multi-controlled NOT gate, we check if the input state is :math:`|i\rangle` and
+        store the information in the auxiliary qubit (if the state is :math:`|i\rangle` the auxiliary
+        qubit will be in the :math:`|1\rangle` state).
+
+        2. If the auxiliary qubit is in the :math:`|1\rangle` state, we modify the input state by adding
+        PauliX gates in the bits that are different between :math:`|i\rangle` and :math:`|b_i\rangle`.
+
+        3. By using a multi-controlled NOT gate, we check if the final state is :math:`|b_i\rangle` and
+        return to :math:`|0\rangle` the auxiliary qubit.
+
+        Appliying these two blocks together results in the desired superposition:
+
+        .. math::
+
+            |\phi\rangle = \sum_i^m c_i |b_i\rangle.
+
+        The decomposition has a complexity that grows linearly with the number of terms,
+        unlike other methods such as :class:`~.MottonenStatePreparation`, that grows exponentially
+        with the number of qubits.
+    """
+
+    num_wires = AnyWires
+    grad_method = None
+    ndim_params = (1,)
+
+    def __init__(
+        self, coeffs, basis, wires, work_wire, id=None
+    ):  # pylint: disable=too-many-positional-arguments, too-many-arguments
+
+        unique_basis = qml.math.unique(
+            qml.math.array([tuple(sub_array) for sub_array in basis]), axis=0
+        )
+
```

Comment:
Please put checks to make sure:
* the elements of the basis states cannot be any values other than 0 and 1
* the input superposition is normalised (if it has to be according to the decomposition)
* the basis states have the same length, so the user cannot put [[0, 1], [1, 0, 0]]